common/unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java

```diff
@@ -403,6 +403,44 @@ public void split() {
       new UTF8String[]{fromString("ab"), fromString("def,ghi,")}));
   }
 
+  @Test
+  public void replace() {
+    assertEquals(
+      fromString("re123ace"),
+      fromString("replace").replace(fromString("pl"), fromString("123")));
+    assertEquals(
+      fromString("reace"),
+      fromString("replace").replace(fromString("pl"), fromString("")));
+    assertEquals(
+      fromString("replace"),
+      fromString("replace").replace(fromString(""), fromString("123")));
```

Comment:
I realized that Apache Commons StringUtils differs from Java `.replace()` in the special case where the search sring is empty: In Java, this would prepend the replacement string to each character, e.g. `"ABC.replace("", 1")` is "1A1B1C".

I also spotted a case where `UTF8String` diverges from Java strings: in Java, we have

```
@ "".replace("", "!")
res2: String = "!"
```

but the existing Spark `UTF8String.replace()` always returns an empty string when passed an empty input string. I'm going to preserve that existing behavior.

Reply:
**Update**: I confused myself andI had this backwards:

 I ran a couple of tests in `spark-shell` with the SQL replace expression:

```
%sql select replace('ABC', '', '1')
ABC
--

because the old Spark code had a `if (EMPTY_UTF8.equals(search)) { return this }` branch (so Spark's behavior matches StringUtils in this case.

If you don't permit search strings then also it's safe to bail out early if the input string is empty (since an empty input string can only become non-empty via an empty search string, something we already disallowed).

As a result, I think that the logic here is correct. Sorry for the false-alarm!

Reply:
In fact, this behavior is also tested at https://github.com/apache/spark/blob/fe5145ede2be50185d77ad3843e95be060d4a73e/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala#L415 and I didn't change those tests in this PR, so we can be extra confident that the new short-circuit check is correct.